### PR TITLE
Always write headers and check against empty paths

### LINKED ISSUE #78: Not specifying output header file (-h) casues ArgumentException
In PInvokeGeneratorConfiguration, headerFile seems to be null-checked and passed to `File.ReadAllText`:

https://github.com/microsoft/ClangSharp/blob/b15d4a79facfec91782abfec91236fe031c2c9df/sources/ClangSharp.PInvokeGenerator/PInvokeGeneratorConfiguration.cs#L54

However, the headerFile option defaults to `string.Empty`, meaning this check always succeeds:

https://github.com/microsoft/ClangSharp/blob/b15d4a79facfec91782abfec91236fe031c2c9df/sources/ClangSharpPInvokeGenerator/Program.cs#L329-L337

This causes the following exception if the parameter is not specified, when running the program:
```
System.ArgumentException: Empty path name is not legal.
Parameter name: path
   at System.IO.File.ReadAllText(String path)
   at ClangSharp.PInvokeGeneratorConfiguration..ctor(String libraryPath, String namespaceName, String outputLocation, PInvokeGeneratorConfigurationOptions options, String[] excludedNames, String headerFile, String methodClassName, String methodPrefixToStrip, IReadOnlyDictionary`2 remappedNames, String[] traversalNames) in /home/dev/Development/ClangSharp/sources/ClangSharp.PInvokeGenerator/PInvokeGeneratorConfiguration.cs:line 54
   at ClangSharp.Program.Run(InvocationContext context) in /home/dev/Development/ClangSharp/sources/ClangSharpPInvokeGenerator/Program.cs:line 173
```

### LINKED ISSUE #79: Output header file is not used for generated types in multi-file mode
When using `--config multi-file`, ClangSharpPInvokeGenerator does not use the header file (`-h`) when generating files for subtypes - only the main methods file gets the header file.

From cursory inspection, this is probably because the generated subtypes do not have any usings, and the header text is only added if there are any using directives:

https://github.com/microsoft/ClangSharp/blob/b15d4a79facfec91782abfec91236fe031c2c9df/sources/ClangSharp.PInvokeGenerator/PInvokeGenerator.cs#L271-L273

### PR DESCRIPTION
**Pull request by FiniteReality**
This fixes #78 and fixes #79. Header files are written whenever a valid header file is passed, and null/empty paths are now ignored.

Sorry about the whitespace changes, I didn't realise I had VS Code set to hide whitespace changes in git diffs.

### CODE REVIEW

File: sources/ClangSharp.PInvokeGenerator/PInvokeGenerator.cs [hunk 1]
```diff
@@ -1399,12 +1400,12 @@ private void VisitParmVarDecl(ParmVarDecl parmVarDecl, TypedefDecl typedefDecl)
             var parameters = typedefDecl.CursorChildren.Where((cursor) => cursor is ParmVarDecl).Cast<ParmVarDecl>().ToList();
             var index = parameters.IndexOf(parmVarDecl);
             var lastIndex = parameters.Count - 1;
-            
+
             if (name.Equals("param"))
             {
                 _outputBuilder.Write(index);
             }
-
```
Comment (tannergooding) [inline sources/ClangSharp.PInvokeGenerator/PInvokeGenerator.cs]: 👍, the formatter should have removed this and it shouldn't have been committed in the first place

File: sources/ClangSharp.PInvokeGenerator/PInvokeGenerator.cs [hunk 2]
```diff
@@ -268,10 +268,11 @@ private void CloseOutputBuilder(Stream stream, OutputBuilder outputBuilder, bool
             using var sw = new StreamWriter(stream, defaultStreamWriterEncoding, DefaultStreamWriterBufferSize, leaveStreamOpen);
             sw.NewLine = "\n";
 
+            if (_config.HeaderText != string.Empty && !isMethodClass)
```
Comment (tannergooding) [inline sources/ClangSharp.PInvokeGenerator/PInvokeGenerator.cs]: Where will it be handled for the `methodClass` if not here?

Comment (tannergooding) [reply sources/ClangSharp.PInvokeGenerator/PInvokeGenerator.cs]: Also, won't https://github.com/microsoft/ClangSharp/issues/79 also apply to the single-file case if you have a header file without methods? So shouldn't https://github.com/microsoft/ClangSharp/blob/master/sources/ClangSharp.PInvokeGenerator/PInvokeGenerator.cs#L86 be fixed too?

Something like:
```cs
                using var sw = new StreamWriter(stream, defaultStreamWriterEncoding, DefaultStreamWriterBufferSize, leaveStreamOpen);
                {
                    sw.NewLine = "\n";
                    sw.Write(_config.HeaderText);

                    foreach (var usingDirective in usingDirectives)
                    {
                        sw.Write("using");
                        sw.Write(' ');
                        sw.Write(usingDirective);
                        sw.WriteLine(';');
                    }

                    sw.WriteLine();
                }
```

Comment (FiniteReality) [reply sources/ClangSharp.PInvokeGenerator/PInvokeGenerator.cs]: From my tests this doesn't seem to happen, and the header always ends up getting written. I am not completely sure if I have tested every edge case though.

Maybe a better approach would be to change CloseOutputBuilder to accept an "additional usings" parameter and have it write everything, including the header?

Comment (tannergooding) [reply sources/ClangSharp.PInvokeGenerator/PInvokeGenerator.cs]: Even in the case where there are no using statements for the single file?

Comment (tannergooding) [reply sources/ClangSharp.PInvokeGenerator/PInvokeGenerator.cs]: Whats the output in that case?

Say if, for example, you just have a header file containing:
```cs
struct MyStruct
{
    int value;
}
```

**Comment by tannergooding** [comment]
Test failures for windows look unrelated. Looks like something might've changed with how paths are processed....

File: sources/ClangSharp.PInvokeGenerator/PInvokeGenerator.cs (hunk 2 above)
Comment (tannergooding) [reply sources/ClangSharp.PInvokeGenerator/PInvokeGenerator.cs]: Ok, looks like the exact case above is fine, but if you do:
```
#include <stdint.h>

struct MyStruct
{
    intptr_t value;
};
```

Then you get the license file printed twice.

Comment (tannergooding) [reply sources/ClangSharp.PInvokeGenerator/PInvokeGenerator.cs]:
```cs
// Copyright
using System;

// Copyright

namespace LLVMSharp
{
    public partial struct MyStruct
    {
        [NativeTypeName("intptr_t")]
        public IntPtr value;
    }
}
```

Comment (tannergooding) [reply sources/ClangSharp.PInvokeGenerator/PInvokeGenerator.cs]: This should probably be...
```cs
if (_config.GenerateMultipleFiles)
{
    if (_config.HeaderText != string.Empty)
    {
        sw.WriteLine(_config.HeaderText);
    }

    if (outputBuilder.UsingDirectives.Any()
    {
        ...
    }
}
```

Comment (tannergooding) [reply sources/ClangSharp.PInvokeGenerator/PInvokeGenerator.cs]: (The `!isMethodClass` was also causing issues, as the license wasn't generating for that file)

Comment (FiniteReality) [reply sources/ClangSharp.PInvokeGenerator/PInvokeGenerator.cs]: :+1:

**Comment by tannergooding** [comment]
Thanks @FiniteReality
